### ISSUE TITLE
cuda ufunc unnecessarily calling `__cuda_array_interface__` 

### ISSUE THREAD

**Issue by sklam** [numba/numba#5191]
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

cuda ufunc is unnecessarily calling `__cuda_array_interface__` and it's increasing call overhead.

The problem can be observed in:

```python
from numba import vectorize, cuda

import numpy as np

a=cuda.to_device(np.ones(10))

@vectorize("float64(float64)", target='cuda')
def foo(x):
    return x

foo(a)
```

`@cuda.jit` path is unaffected.

**Comment by gmarkall** [comment]
Perhaps this matters less now that #5189 is merged?

**Comment by cobalamin** [comment]
I can confirm a performance regression in 0.48 that is gone in 0.49.0rc1 even with a very simple function (taken from chapter 3 in https://github.com/ContinuumIO/gtc2017-numba. In the more current https://github.com/numba/euroscipy2019-numba, the code for this part is essentially the same):

### Code used
N=10^6, 10^7 or 10^8.

```python
from numba import vectorize, cuda
import numpy as np

@vectorize(['float32(float32, float32)'], target='cuda')
def add_ufunc(x, y):
    return x + y
```

```python
# (1)
x = np.arange(N).astype(np.float32)
y = 2 * x

%timeit add_ufunc(x, y)  # Baseline performance with host arrays
```

```python
# (2)
x_device = cuda.to_device(x)
y_device = cuda.to_device(y)

%timeit add_ufunc(x_device, y_device)
```

```python
# (3)
out_device = cuda.device_array(shape=(n,), dtype=np.float32)

%timeit add_ufunc(x_device, y_device, out=out_device)
```

### Results

| N, version // cell   | (1)           | (2)           | (3)           | (1)/(3) |
| -------------------- | ------------- | ------------- | ------------- | ------- |
| 10^5, 0.48           | 831us         | 748us         | 923us         | 0.90    |
| 10^6, 0.48           | 2.06ms        | 937us         | 916us         | 2.25    |
| 10^7, 0.48           | 16.2ms        | 1.64ms        | 924us         | 17.53   |
| -------------------- | ------------- | ------------- | ------------- | ------- |
| 10^5, 0.49.0rc1      | 852us         | 373us         | 330us         | 2.58    |
| 10^6, 0.49.0rc1      | 2.1ms         | 537us         | 337us         | 6.23    |
| 10^7, 0.49.0rc1      | 16.2s         | 1.29ms        | 338us         | 47.93   |

**Comment by esc** [comment]
@cobalamin thanks for reporting back about this, good to hear that you are getting these numbers now.

**Comment by jakirkham** [comment]
So then is this already fixed?

**Comment by gmarkall** [comment]
Ufuncs probably are still calling `__cuda_array_interface__` unnecessarily, but doing so is much quicker than it was before due to https://github.com/numba/numba/pull/5189. So I'm not sure whether to consider this "fixed" or not.

**Comment by gmarkall** [comment]
This will be fixed when #5162 is merged.
